src/puppetlabs/puppetdb/cli/services.clj

```diff
@@ -422,14 +441,16 @@
   that trapperkeeper will call on exit."
   PuppetDBServer
   [[:DefaultedConfig get-config]
-   [:WebroutingService add-ring-handler get-registered-endpoints]]
+   [:WebroutingService add-ring-handler get-registered-endpoints]
+   [:ShutdownService shutdown-on-error]]
   (init [this context]
 
         (doseq [{:keys [reporter]} (vals metrics/metrics-registries)]
           (jmx-reporter/start reporter))
         (assoc context :url-prefix (atom nil)))
   (start [this context]
-         (start-puppetdb context (get-config) this get-registered-endpoints))
+         (start-puppetdb context (get-config) this get-registered-endpoints
+                         (partial shutdown-on-error (service-id this))))
```

Comment:
This seems like a pretty idiomatic place to use exceptions. Instead of passing down this 'shutdown-on-error' function, we could have initial-schema check the database version as a precondition and `(throw+ {:type ::unsupported-database, :required-db-version [9 6], :found-pg-version [9 4]}`, then `(try+ ... (catch [:type :scf/unsupported-database] (log blah) (shutdown blarg)))` here at the top level.